### DISCUSSION TITLE
TemporaryUploadedFile::temporaryUrl() should return fallback instead of exception

### DISCUSSION THREAD

**Discussion by michaelkonstantinou** [livewire/livewire#7269] [3. Feature ideas]
Hi, I had a while ago found an issue with `temporaryUrl` throwing exceptions (often unrelated to the driver) and initially I thought that perhaps I could bypass it by adding a function that catches the exception. Long story short, it got me thinking on how this could be fixed and I have a solution that I would be more than happy to open a PR for it, if the community also agrees.

**Problem**: `TemporaryUploadedFile::temporaryUrl()`:

- is used mainly in front-end
- throws an exception
- the exception thrown might be irrelevant to the driver... yet it throws a driver exception
- there is an inconsistency between how the filesystem class checks whether `temporaryUrl` is supported and how `TemporaryUploadedFile` checks whether `temporaryUrl` is supported by the driver

**Solution**: `TemporaryUploadedFile::temporaryUrl()` must change to:

- Have better rules
- Instead of throwing an exception (which is dangerous when calling it in the front-end), the function could provide a fallback url.

### Some issues with the current functionality
The "inconsistency" in the storage's temporaryUrl and livewire's way of checking whether it can call it or not

**FilesystemAdapter::providesTemporaryUrls**
```php
   /**
   * Determine if temporary URLs can be generated.
   *
   * @return bool
   */
  public function providesTemporaryUrls()
  {
      return method_exists($this->adapter, 'getTemporaryUrl') || isset($this->temporaryUrlCallback);
  }
```

**TemporaryUploadedFile::temporaryUrl**
```php
if ((FileUploadConfiguration::isUsingS3() or FileUploadConfiguration::isUsingGCS()) && ! app()->runningUnitTests()) {
            // returns something that does not break
        }

        if (method_exists($this->storage->getAdapter(), 'getTemporaryUrl') || ! $this->isPreviewable()) {
            // This will throw an error because it's not used with S3. **But haven't we checked this above?**
            return $this->storage->temporaryUrl($this->path, now()->addDay());
        }
```

- FilesystemAdapter has the function `providesTemporaryUrls` to check whether a temporaryUrl can be generated
- TemporaryUploadedFile calls the function temporaryUrl in method TemporaryUploadedFile::temporaryUrl() after it checks whether the driver is S3 or GCS, and the following returns true: `method_exists($this->storage->getAdapter(), 'getTemporaryUrl') || ! $this->isPreviewable()`
- Which means that while the function FilesystemAdapter::providesTemporaryUrls might return false.... TemporaryUploadedFile still calls the function temporaryUrl

So in practice, this means the following:
- If you do not use S3 or GCS and you have a **valid** **path**, `TemporaryUploadedFile::temporaryUrl` will use `URL::temporarySignedRoute` to create a `temporaryUrl`
- If for any reason the path is changed and is **no longer previewable**... then the **same** `TemporaryUploadedFile` object will use `FilesystemAdapter::temporaryUrl` to create a `temporaryUrl`... which will crash

Usually this happens when a file is uploaded. Once the file is uploaded the path is changed. The solution I do and I see other people doing, is to immediately make the variable `$photo` null. Which still, doesn't mean that this is ideal but it works. In one of the examples I mentioned above the function `is_object` is used.

But this is not the only way to get an exception. If for any reason the path is not recognised as previewable, maybe because a third-party image library crashed, or changed something in the path, then you will end up with a TemporaryUploadedFile that throws an exception. 

TLDR; there is a difference between `FilesystemAdapter::providesTemporaryUrls` and how `TemporaryUploadedFile::temporaryUrl` calls the code `FilesystemAdapter::temporaryUrl` which leaves a small gap to throw an exception. 

### Possible solution

The function should change as such
```
public function temporaryUrl(string $fallbackUrl = '')
    {
        if ((FileUploadConfiguration::isUsingS3() or FileUploadConfiguration::isUsingGCS()) && ! app()->runningUnitTests()) {
            return $this->storage->temporaryUrl(
                $this->path,
                now()->addDay()->endOfHour(),
                ['ResponseContentDisposition' => 'attachment; filename="' . $this->getClientOriginalName() . '"']
            );
        }

        if ($this->storage->providesTemporaryUrls()) {
            // This will NOT throw an error anymore
            return $this->storage->temporaryUrl($this->path, now()->addDay());
        }

       if ($this->isPreviewable()) {
            return URL::temporarySignedRoute(
            'livewire.preview-file', now()->addMinutes(30)->endOfHour(), ['filename' => $this->getFilename()]
        );
       }

       return $fallbackUrl;
    }
```

### TLDR questions:

- What is the issue with the current implementation?
- - The current implementation throws an exception because its rules are not properly defined. It may throw a driver exception when in reality the issue was that the image was damaged or "not previewable"
-  Why fallback url?
- - Fallback url ensures that the error will be handled in a user-friendly way. Image the issue when a popup 500 shows or the image's alt text is displayed. Now imagine how more user-friendly would be to show a gray image telling the user that the image is not previewable just like youtube does

Let me know what you think. If you all agree I can open a PR for this issue and have a better generation of temporary urls
